### PR TITLE
Add attribute validator.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

**Review by samdark — APPROVED** [review]

File: src/Validator.php [hunk 1]
```diff
@@ -34,10 +35,21 @@ public function __construct(private RuleHandlerResolverInterface $ruleHandlerRes
     public function validate(mixed $data, ?iterable $rules = null): Result
     {
         $data = $this->normalizeDataSet($data);
+
+        $rulesByAttribute = [];
+
+        if ($data instanceof DataSetInterface) {
```
Comment (samdark) [inline src/Validator.php:41]: Attributes are supported in non-DataSetInterface.

Comment (samdark) [reply src/Validator.php:41]: The check should be is_object.

Comment (terabytesoftw) [reply src/Validator.php]: done.

**Comment by terabytesoftw** [comment]
This PR allows using the rules by `attribute` and defined in `getRules()` simultaneously, where the developer has the ease of using simple rules defining them as `attributes`, and complex rules and `callbacks` or `closures` as definitions in `getRuleS()`.

File: src/Validator.php [hunk 2]
```diff
@@ -34,8 +35,11 @@ public function __construct(private RuleHandlerResolverInterface $ruleHandlerRes
     public function validate(mixed $data, ?iterable $rules = null): Result
     {
         $data = $this->normalizeDataSet($data);
+
         if ($rules === null && $data instanceof RulesProviderInterface) {
```
Comment (xepozz) [inline src/Validator.php:39]: could you also change validator collector in the debug package?

Comment (terabytesoftw) [reply src/Validator.php:39]: done.

**Comment by terabytesoftw** [comment]
https://github.com/yiisoft/yii-debug/pull/146

**Review by xepozz — APPROVED** [review]

File: src/Validator.php [hunk 3]
```diff
@@ -36,7 +36,9 @@ public function validate(mixed $data, ?iterable $rules = null): Result
     {
         $data = $this->normalizeDataSet($data, $rules !== null);
         if ($rules === null && $data instanceof RulesProviderInterface) {
-            $rules = $data->getRules();
+            $rules = (array) $data->getRules();
+            $rulesByAttribute = (array) ((new AttributeDataSet($data))->getRules());
```
Comment (samdark) [inline src/Validator.php:40]: Casting to array doesn't work well for all possible iterables.

**Comment by samdark** [comment]
It seems to make https://github.com/yiisoft/validator/issues/261 useless. @xepozz, @terabytesoftw  am I right?

**Comment by samdark** [comment]
Could be replaced by https://github.com/yiisoft/validator/pull/265

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -23,15 +23,10 @@ name: build
...
+    uses: yiisoft/actions/.github/workflows/phpunit.yml@master
     with:
       extensions: intl
-      os: &phpunit-os >-
+      os: >-
         ['ubuntu-latest', 'windows-latest']
-      php: &phpunit-php >-
+      php: >-
         ['8.0', '8.1']
-  phpunit_without_extensions:
```
Comment (arogachev) [inline .github/workflows/build.yml:33]: Regression for #277.

**Comment by arogachev** [comment]
What about tests, docs?

**Review by arogachev — CHANGES_REQUESTED** [review]
What about tests, docs?

**Comment by xepozz** [comment]
> It seems to make https://github.com/yiisoft/validator/issues/261 useless. @xepozz, @terabytesoftw  am I right?

Yep, but we can solve it later

**Comment by terabytesoftw** [comment]
Closed in favor #261

**Pull request closed without merging**